pkg/csi/manila/controllerserver.go

```diff
@@ -332,7 +340,58 @@ func (cs *controllerServer) ControllerGetCapabilities(ctx context.Context, req *
 }
 
 func (cs *controllerServer) ValidateVolumeCapabilities(ctx context.Context, req *csi.ValidateVolumeCapabilitiesRequest) (*csi.ValidateVolumeCapabilitiesResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	if err := validateValidateVolumeCapabilitiesRequest(req); err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	osOpts, err := options.NewOpenstackOptions(req.GetSecrets())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid OpenStack secrets: %v", err)
+	}
+
+	for _, volCap := range req.GetVolumeCapabilities() {
+		if volCap.GetBlock() != nil {
+			return &csi.ValidateVolumeCapabilitiesResponse{Message: "block access type is not allowed"}, nil
+		}
+
+		if volCap.GetMount() == nil {
+			return &csi.ValidateVolumeCapabilitiesResponse{Message: "volume must be accessible via filesystem API"}, nil
+		}
+
+		if volCap.GetAccessMode().GetMode() == csi.VolumeCapability_AccessMode_UNKNOWN {
+			return &csi.ValidateVolumeCapabilitiesResponse{Message: "unknown volume access mode"}, nil
+		}
+	}
+
+	manilaClient, err := cs.d.manilaClientBuilder.New(osOpts)
+	if err != nil {
+		return nil, status.Errorf(codes.Unauthenticated, "failed to create Manila v2 client: %v", err)
+	}
+
+	share, err := manilaClient.GetShareByID(req.GetVolumeId())
+	if err != nil {
+		if isManilaErrNotFound(err) {
+			return nil, status.Errorf(codes.NotFound, "share %s not found: %v", req.GetVolumeId(), err)
+		}
+
+		return nil, status.Errorf(codes.Internal, "failed to retrieve share %s: %v", req.GetVolumeId(), err)
+	}
+
+	if share.Status != shareAvailable {
+		return nil, status.Errorf(codes.InvalidArgument, "share %s is in an unexpected state: wanted %s, got %s", share.ID, shareAvailable, share.Status)
```

Comment:
here also, shouldn't it be codes.Internal?

Reply:
InvalidArgument, is sent when something is missing in request from CO or request arguments not adhering to CSI specs

Reply:
Actually, `FAILED_PRECONDITION` or `UNAVAILABLE` seem to come even closer to describing this particular error. The status mismatch might be just a transient error (i.e. when the share is still in `creating` state).

https://github.com/grpc/grpc-go/blob/3bb34e5324978dca98dea479160fe5f434d0f960/codes/codes.go#L78-L97